File: 2015/23/Lua/main.lua
-- Function to split a string by a delimiter
local function split_string(input)
    local res = {}
    for match in input:gmatch("%S+") do
        table.insert(res, match)
    end
    return res
end

-- Read and split file contents
local function read_file_into_table(path)
    local file = io.open(path, "r")
    local instructions = {}
    if not file then
        return nil, "Error opening file."
    end

    for line in file:lines() do
        line = line:gsub(",", "")
        line = split_string(line)
        table.insert(instructions, line)
    end
    file:close()
    return instructions
end

local function apply_instruction(instruction, register_val, position, offset)
    if instruction == "hlf" then
        register_val = register_val // 2
        position = position + 1
    elseif instruction == "tpl" then
        register_val = register_val * 3
        position = position + 1
    elseif instruction == "inc" then
        register_val = register_val + 1
        position = position + 1
    elseif instruction == "jie" then
        position = (register_val % 2 == 0) and (position + offset) or (position + 1)
    elseif instruction == "jio" then
        position = (register_val == 1) and (position + offset) or (position + 1)
    end
    return register_val, position
end

local function exec_instructions(instructions, registers)
    local i = 1
    while i>0 and i <= #instructions do
        local instruction, action, offset = table.unpack(instructions[i])
        offset = tonumber(offset) or 0
        if instruction == "jmp" then
            i = i + action
        else
            registers[action], i = apply_instruction(instruction, registers[action], i, offset)
        end
    end
    return registers
end

local function main(path)
    local instructions, err = read_file_into_table(path)
    if instructions then
        local registers = exec_instructions(instructions, { a = 0, b = 0 })
        print("First challenge: the value for register b is " .. registers["b"])

        registers = exec_instructions(instructions, { a = 1, b = 0 })
        print("Second challenge: the value for register b is " .. registers["b"])
    else
        error(string.format("Error reading file %s: %s", path, err))
    end
end

local path = "input.txt"
main(path)

File: 2015/23/Go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type InstructionLine struct {
	instruction string
	register    string
	offset      int
}

func parseLine(line string) (InstructionLine, error) {
	line = strings.TrimSpace(strings.ReplaceAll(line, ",", ""))
	parts := strings.Split(line, " ")
	instr := InstructionLine{instruction: parts[0]}
	// Jmp instruction is different
	var offset string = "0"
	if parts[0] == "jmp" {
		offset = parts[1]
	} else { // Any other instruction
		instr.register = parts[1]
		if len(parts) == 3 {
			offset = parts[2]
		}
	}
	parsedOffset, err := strconv.Atoi(offset)
	if err != nil {
		return InstructionLine{}, fmt.Errorf("error parsing offset %s: %v", offset, err)
	}
	instr.offset = parsedOffset
	return instr, nil
}

func applyInstruction(line InstructionLine, registerVal int, position int) (int, int) {
	switch line.instruction {
	case "hlf":
		registerVal /= 2
	case "tpl":
		registerVal *= 3
	case "inc":
		registerVal++
	case "jie":
		if registerVal%2 == 0 {
			position += line.offset
			return registerVal, position
		}
	case "jio":
		if registerVal == 1 {
			position += line.offset
			return registerVal, position
		}
	}
	position++
	return registerVal, position
}

func execInstructions(instructions []InstructionLine, registers map[string]int) map[string]int {
	i := 0
	for i >= 0 && i < len(instructions) {
		currentInstruction := instructions[i]
		if currentInstruction.instruction == "jmp" {
			i += currentInstruction.offset
			continue
		}
		registers[currentInstruction.register], i = applyInstruction(currentInstruction,
			registers[currentInstruction.register], i)
	}
	return registers
}

func readFile(filepath string) ([]InstructionLine, error) {
	file, err := os.Open(filepath)
	if err != nil {
		return nil, fmt.Errorf("opening file %s: %v", filepath, err)
	}
	defer file.Close()

	var instructions = []InstructionLine{}
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		parsed, err := parseLine(scanner.Text())
		if err != nil {
			return nil, fmt.Errorf("unable to parse line %v", err)
		}
		instructions = append(instructions, parsed)
	}
	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("scanning file %s: %v", filepath, err)
	}
	return instructions, nil
}

func main() {
	filepath := "input.txt"
	registers := make(map[string]int, 2)
	instructions, err := readFile(filepath)
	if err != nil {
		fmt.Printf("Unable to read the file %s:\n %v\n", filepath, err)
		return
	}
	registers["a"], registers["b"] = 0, 0
	registers = execInstructions(instructions, registers)
	fmt.Println("First challenge: the value for register b is", registers["b"])

	registers["a"], registers["b"] = 1, 0
	registers = execInstructions(instructions, registers)
	fmt.Println("Second challenge: the value for register b is", registers["b"])
}
